src/azul/deployment.py

```diff
@@ -434,8 +437,46 @@ def client(self, *args, **kwargs):
         Caching the result of this function is not necessary and will be harmful
         if the cached value is used by a thread other than the one that called
         this function.
+
+        :param azul_logging: whether to log the client's requests and responses
+        """
+        client = self.boto3_session.client(*args, **kwargs)
+        if azul_logging:
+            client.meta.events.register('before-send', self._log_client_request)
+            client.meta.events.register('response-received', self._log_client_response)
+        return client
+
+    def _log_client_request(self,
+                            event_name: str,
+                            request: AWSPreparedRequest,
+                            **kwargs: Any
+                            ) -> Optional[AWSResponse]:
+        """
+        Log the request's AWS API action, HTTP method, URL, and body (unless
+        the body is streamable).
+        """
+        event_name = event_name.replace('before-send', 'client')
+        log.info('%s requesting %s to %s', event_name, request.method, request.url)
+        if isinstance(request.body, (str, bytes, bytearray)):
+            log.info('… with request body %r', trunc_ellipses(request.body, 128))
+        return None
+
+    def _log_client_response(self,
+                             *,
+                             event_name: str,
+                             **kwargs: Any
+                             ) -> Optional[AWSResponse]:
+        """
+        Log the request's AWS API action and the response's HTTP status and
+        body (unless the body is streamable).
         """
-        return self.boto3_session.client(*args, **kwargs)
+        event_name = event_name.replace('response-received', 'client')
+        response = kwargs.get('response_dict')
```

Comment:
If `get` retruns `None`, then the next line will raise an `AttributeError`, so this should be either
```suggestion
        response = kwargs.get('response_dict', {})
```
or
```suggestion
        response = kwargs['response_dict']
```